development/orphan-remover/main.go

```diff
@@ -0,0 +1,331 @@
+
+import (
+	"context"
+	// "encoding/json"
+	"flag"
+	"fmt"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	compute "google.golang.org/api/compute/v1"
+	"log"
+	"os"
+	"strings"
+	"time"
+)
+
+var (
+	project         = flag.String("project", "", "Project ID")
+	dryRun          = flag.Bool("dry-run", true, "Dry Run enabled")
+	targetPool      = []TargetPool{}
+	garbagePool     = []TargetPool{}
+	instanceGroups  = []InstanceGroup{}
+	backendServices = []BackendService{}
+	urlMaps         = []URLMap{}
+)
+
+//TargetPool ???
+type TargetPool struct {
+	name          string
+	instances     []Instance
+	instanceCount int
+	healthChecks  []string
+	region        string
+	timestamp     string
+}
+
+// Instance ???
+type Instance struct {
+	name   string
+	zone   string
+	exists bool
+}
+
+// InstanceGroup ???
+type InstanceGroup struct {
+	name string
+	id   string
+	zone string
+}
+
+// BackendService ???
+type BackendService struct {
+	name string
+	id   string
+}
+
+// URLMap ???
+type URLMap struct {
+	name string
+	id   string
+}
+
+func main() {
+	flag.Parse()
+
+	if *project == "" {
+		fmt.Fprintln(os.Stderr, "missing -project flag")
+		flag.Usage()
+		os.Exit(2)
+	}
+
+	context := context.Background()
+	connenction, err := google.DefaultClient(context, compute.CloudPlatformScope)
+	if err != nil {
+		log.Fatalf("Could not get authenticated client: %v", err)
+	}
+
+	svc, err := compute.New(connenction)
+	if err != nil {
+		log.Fatalf("Could not initialize gke client: %v", err)
+	}
+
+	targetPool, _ := lookupTargetPools(svc, *project)
+	for _, target := range targetPool {
+		markCount := 0
+		for _, instance := range target.instances {
+			markInstance(svc, *project, &instance)
+			if !instance.exists {
+				markCount++
+			}
+		}
+		if markCount == target.instanceCount {
+			garbagePool = append(garbagePool, target)
+		}
+	}
+
+	fmt.Printf("All items: %d\n", len(targetPool))
+	fmt.Printf("Garbage items: %d\n", len(garbagePool))
+
+	zones, _ := lookupZones(svc, *project, "europe-*")
+	for _, zone := range zones {
+		igList, _ := lookupInstanceGroup(svc, *project, zone)
+		if len(igList) > 0 {
+			for _, name := range igList {
+				fields := strings.Split(name, "--")
+				id := fields[len(fields)-1]
+				instanceGroups = append(instanceGroups, InstanceGroup{name, id, zone})
+			}
+		}
+	}
+	urlMaps, _ := lookupURLMaps(svc, *project)
+	backendServices, _ := lookupBackendServices(svc, *project)
+	purge(svc, garbagePool, instanceGroups, backendServices, urlMaps, *dryRun)
+}
+
+func purge(svc *compute.Service, targetPool []TargetPool, instanceGroups []InstanceGroup, backendServices []BackendService, urlMaps []URLMap, dryRun bool) {
+	for _, target := range targetPool {
+		fmt.Printf("-> Processing targetPool: %s\n", target.name)
+
+		fmt.Printf("---> Delete ForwardingRules: %s in Region: %s\n", target.name, target.region)
+		if !dryRun {
+			deleteForwardingRule(svc, *project, target.name, target.region)
+			time.Sleep(5 * time.Second)
```

Comment:
why sleeps here?

Reply:
Sleeps are required as deleting a resource takes some time, but the api command just sends the request and does not wait for confirmation or completion. Since we are deleting connected resources, we need to either wait for some time, or would need to ask the api if the resource has already been deleted before we delete another.